### PR TITLE
API Add change password functionality via LDAP server

### PR DESCRIPTION
**Pull request by robbieaverill**
## Requirements

- [x] https://github.com/silverstripe/silverstripe-framework/pull/7428
- [x] https://github.com/silverstripe/silverstripe-framework/pull/7424

## Changes

* Fix PSR-4 autoloading in test classes
* Remove overloaded logic from core login handlers (requires https://github.com/silverstripe/silverstripe-framework/pull/7428)
* Move "update the password in LDAP when it changes in SilverStripe" to an extension rather than more overloaded logic (requires https://github.com/silverstripe/silverstripe-framework/pull/7424)
* Add "check password" support to the LDAP authenticator, required for resetting your password
* Remove unused LDAPMemberLoginHandler class
* Fix CI bootstrapping paths

### CODE REVIEW

File: docs/en/developer.md [hunk 1]
```diff
@@ -511,10 +519,12 @@ that the "Username" field must be filled in, otherwise it will not be created, d
...
-    $member = new \SilverStripe\Security\Member();
-    $member->FirstName = 'Joe';
-    $member->Username = 'jbloggs';
-    $member->write();
+```php
+$member = new \SilverStripe\Security\Member();
+$member->FirstName = 'Joe';
+$member->Username = 'jbloggs';
+$member->write();
+```
```
Comment (robbieaverill) [inline docs/en/developer.md:527]: More extensive docs rewrite is coming next *edit* #6

**Comment by robbieaverill** [comment]
SS core doesn't trigger `Member::changePassword` when you update a password in a member profile via the admin interface. It would be ideal if it did, but it's only being used in the "forgot my password" type login forms at the moment. @halkyon @mateusz what are your thoughts on this? Log a separate issue/PR to core to make it use `changePassword`/add a hack to this module instead?

*Edit:* I've logged this on framework for now - https://github.com/silverstripe/silverstripe-framework/issues/7435

**Review by NightJar — APPROVED** [review]

File: src/Authenticators/LDAPAuthenticator.php [hunk 2]
```diff
@@ -219,4 +219,24 @@ public function getChangePasswordHandler($link)
     {
         return LDAPChangePasswordHandler::create($link, $this);
     }
+
+    public function checkPassword(Member $member, $password, ValidationResult &$result = null)
+    {
+        $result = $result ?: ValidationResult::create();
+
+        /** @var LDAPService $service */
+        $service = Injector::inst()->get(LDAPService::class);
```
Comment (ichaber) [inline src/Authenticators/LDAPAuthenticator.php:228]: We could replace that via Setter Dependency Injection to get the server and allow a custom LDAP Service.

File: tests/php/Authenticators/LDAPChangePasswordHandlerTest.php [hunk 3]
```diff
@@ -0,0 +1,27 @@
+<?php
+
+namespace SilverStripe\LDAP\Tests\Authenticators;
+
+use SilverStripe\Control\HTTPRequest;
+use SilverStripe\Control\Session;
+use SilverStripe\Core\Injector\Injector;
+use SilverStripe\Dev\SapphireTest;
+use SIlverStripe\LDAP\Authenticators\LDAPAuthenticator;
```
Comment (ichaber) [inline tests/php/Authenticators/LDAPChangePasswordHandlerTest.php:9]: That should be use SilverStripe\LDAP\Authenticators\LDAPAuthenticator; (with a lowercase i)

File: tests/php/Authenticators/LDAPChangePasswordHandlerTest.php [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+use SIlverStripe\LDAP\Authenticators\LDAPChangePasswordHandler;
```
Comment (ichaber) [inline tests/php/Authenticators/LDAPChangePasswordHandlerTest.php:10]: That should be use SilverStripe\LDAP\Authenticators\LDAPChangePasswordHandler; (with a lowercase i)

File: tests/php/Authenticators/LDAPChangePasswordHandlerTest.php [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+use SIlverStripe\LDAP\Forms\LDAPChangePasswordForm;
```
Comment (ichaber) [inline tests/php/Authenticators/LDAPChangePasswordHandlerTest.php:11]: That should be use SilverStripe\LDAP\Forms\LDAPChangePasswordForm; (with a lowercase i)

File: src/Authenticators/LDAPAuthenticator.php (hunk 2 above)
Comment (robbieaverill) [reply src/Authenticators/LDAPAuthenticator.php:228]: Good idea, https://github.com/silverstripe/silverstripe-ldap/issues/7

File: tests/php/Authenticators/LDAPChangePasswordHandlerTest.php (hunk 5 above)
Comment (robbieaverill) [reply tests/php/Authenticators/LDAPChangePasswordHandlerTest.php:11]: https://github.com/silverstripe/silverstripe-ldap/issues/8

**Pull request merged**